pkg/runtime.v2/runtime.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2024 The Kubeflow Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package runtimev2
+
+import (
+	"errors"
+	"maps"
+
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	kueuelr "sigs.k8s.io/kueue/pkg/util/limitrange"
+
+	kubeflowv2 "github.com/kubeflow/training-operator/pkg/apis/kubeflow.org/v2alpha1"
+)
+
+var (
+	errorDifferentGVK  = errors.New("the GroupVersionKinds are different between old and new objects")
+	errorObjectsAreNil = errors.New("old or new objects are nil")
+)
+
+type Info struct {
+	Obj         client.Object
+	Labels      map[string]string
+	PodLabels   map[string]string
+	Annotations map[string]string
+	Policy
+	TotalRequests map[string]TotalResourceRequest
+}
+
+type Policy struct {
+	MLPolicy       *kubeflowv2.MLPolicy
+	PodGroupPolicy *kubeflowv2.PodGroupPolicy
+}
+
+type TotalResourceRequest struct {
+	Replicas    int32
+	PodRequests corev1.ResourceList
+}
+
+type InfoOptions struct {
+	podSpecReplicas []podSpecReplica
+	Policy
+	labels      map[string]string
+	annotations map[string]string
+}
+
+type InfoOption func(options *InfoOptions)
+
+var defaultOptions = InfoOptions{}
+
+type podSpecReplica struct {
+	replicas int32
+	name     string
+	podSpec  corev1.PodSpec
+}
+
+func WithPodSpecReplicas(replicaName string, replicas int32, podSpec corev1.PodSpec) InfoOption {
+	return func(o *InfoOptions) {
+		o.podSpecReplicas = append(o.podSpecReplicas, podSpecReplica{
+			name:     replicaName,
+			replicas: replicas,
+			podSpec:  podSpec,
+		})
+	}
+}
+
+func WithLabels(labels map[string]string) InfoOption {
+	return func(o *InfoOptions) {
+		o.labels = maps.Clone(labels)
+	}
+}
+
+func WithAnnotations(annotations map[string]string) InfoOption {
+	return func(o *InfoOptions) {
+		o.annotations = maps.Clone(annotations)
+	}
+}
+
+func WithPodGroupPolicy(pgPolicy *kubeflowv2.PodGroupPolicy) InfoOption {
+	return func(o *InfoOptions) {
+		o.PodGroupPolicy = pgPolicy
+	}
+}
+
+func WithMLPolicy(mlPolicy *kubeflowv2.MLPolicy) InfoOption {
+	return func(o *InfoOptions) {
+		o.MLPolicy = mlPolicy
+	}
+}
+
+func NewInfo(obj client.Object, opts ...InfoOption) *Info {
+	options := defaultOptions
+	for _, opt := range opts {
+		opt(&options)
+	}
```

Comment:
Do we need default options for Info object ?

Reply:
```suggestion
	options := InfoOptions{}
	for _, opt := range opts {
		opt(&options)
	}
```

Do you recommend this?

Reply:
Maybe, I am trying to understand how are your planning to use InfoOptions in other parts ?
@tenzen-y What are the differences between Info{} and InfoOptions{} ?

Reply:
The `infoOptions` is object to set up the `Info` object.
This approach allows us to dynamically specify the parameters to the `Info`.

When we get rid of the `InfoOptions`, we need to specify all parameters everytime or need to directly pass the `Info` object.

Reply:
This approach is well-known to avid the following function:

```golang
// Even if all parameters is not used, all parameters should be specified.
func Foo(paramA string, paramB int, paramC int32, paramD bool, paramE int64)
```

```golang
// After introduced `infoOptions`.
func Foo(params ...InfoOption)
```

Reply:
I see, so maybe let's name it as InfoOption, not defaultOptions to make it clearer, and since we are not going to have default values for Info object.

Reply:
> we are not going to have default values for Info object.

Actually, this is the default info Option. Here, this means that the default is an empty struct.
So, the below indicates to initialize options as a default values (currently default has empty fields)

```
options := defaultOptions
```

Reply:
After future, we need to consider if we should the default parameters to infoOption for example common labels and annotations.

Reply:
I see, that makes sense.